Mathlib/Algebra/Algebra/Bilinear.lean

```diff
@@ -0,0 +1,271 @@
+  rfl
+#align linear_map.mul_right_to_add_monoid_hom LinearMap.mulRight_toAddMonoid_hom
+
+variable {R}
+
+@[simp]
+theorem mul_apply' (a b : A) : mul R A a b = a * b :=
+  rfl
+#align linear_map.mul_apply' LinearMap.mul_apply'
+
+@[simp]
+theorem mulLeft_apply (a b : A) : mulLeft R a b = a * b :=
+  rfl
+#align linear_map.mul_left_apply LinearMap.mulLeft_apply
+
+@[simp]
+theorem mulRight_apply (a b : A) : mulRight R a b = b * a :=
+  rfl
+#align linear_map.mul_right_apply LinearMap.mulRight_apply
+
+@[simp]
+theorem mulLeftRight_apply (a b x : A) : mulLeftRight R (a, b) x = a * x * b :=
+  rfl
+#align linear_map.mul_left_right_apply LinearMap.mulLeftRight_apply
+
+@[simp]
+theorem mul'_apply {a b : A} : mul' R A (a ⊗ₜ b) = a * b :=
+  rfl
+#align linear_map.mul'_apply LinearMap.mul'_apply
+
+@[simp]
+theorem mulLeft_zero_eq_zero : mulLeft R (0 : A) = 0 :=
+  (mul R A).map_zero
+#align linear_map.mul_left_zero_eq_zero LinearMap.mulLeft_zero_eq_zero
+
+@[simp]
+theorem mulRight_zero_eq_zero : mulRight R (0 : A) = 0 :=
+  (mul R A).flip.map_zero
+#align linear_map.mul_right_zero_eq_zero LinearMap.mulRight_zero_eq_zero
+
+end NonUnitalNonAssoc
+
+section NonUnital
+
+variable (R A : Type _) [CommSemiring R] [NonUnitalSemiring A] [Module R A] [SMulCommClass R A A]
+  [IsScalarTower R A A]
+
+/-- The multiplication in a non-unital algebra is a bilinear map.
+
+A weaker version of this for non-unital non-associative algebras exists as `LinearMap.mul`. -/
+def NonUnitalAlgHom.lmul : A →ₙₐ[R] End R A :=
+  {
+    mul R A with
+    map_mul' := by
+      intro a b
+      ext c
+      exact mul_assoc a b c
+    map_zero' := by
+      ext a
+      exact zero_mul a }
+#align non_unital_alg_hom.lmul LinearMap.NonUnitalAlgHom.lmul
+
+variable {R A}
+
+@[simp]
+theorem NonUnitalAlgHom.coe_lmul_eq_mul : ⇑(NonUnitalAlgHom.lmul R A) = mul R A :=
+  rfl
+#align non_unital_alg_hom.coe_lmul_eq_mul LinearMap.NonUnitalAlgHom.coe_lmul_eq_mul
+
+theorem commute_mulLeft_right (a b : A) : Commute (mulLeft R a) (mulRight R b) := by
+  ext c
+  exact (mul_assoc a c b).symm
+#align linear_map.commute_mul_left_right LinearMap.commute_mulLeft_right
+
+@[simp]
+theorem mulLeft_mul (a b : A) : mulLeft R (a * b) = (mulLeft R a).comp (mulLeft R b) := by
+  ext
+  simp only [mulLeft_apply, comp_apply, mul_assoc]
+#align linear_map.mul_left_mul LinearMap.mulLeft_mul
+
+@[simp]
+theorem mulRight_mul (a b : A) : mulRight R (a * b) = (mulRight R b).comp (mulRight R a) := by
+  ext
+  simp only [mulRight_apply, comp_apply, mul_assoc]
+#align linear_map.mul_right_mul LinearMap.mulRight_mul
+
+end NonUnital
+
+section Semiring
+
+variable (R A : Type _) [CommSemiring R] [Semiring A] [Algebra R A]
+
+/-- The multiplication in an algebra is an algebra homomorphism into the endomorphisms on
+the algebra.
+
+A weaker version of this for non-unital algebras exists as `NonUnitalAlgHom.mul`. -/
+def Algebra.lmul : A →ₐ[R] End R A :=
+  {
+    LinearMap.mul R
```

Comment:
What is the style convention in Mathlib4? Should these two lines just be
```suggestion
  { LinearMap.mul R
```

Reply:
(same comment for `NonUnitalAlgHom.lmul` above)

Reply:
Yeah, that's a bug in mathport. Please merge those lines.